Draft PR: Complex navigator

jmedinaebi opened the pull request:
Just to add comments

susiehgt committed:
Matrix created

The matrix has been created, the component added in it but they are not working with the complexes

susiehgt committed:
Delete complex-results.component.ts

susiehgt committed:
Delete complex-list.component.html

susiehgt committed:
Delete complex-list.component.css

susiehgt committed:
Matrix created

The matrix has been created, the component added in it but they are not working with the complexes

susiehgt committed:
Matrix created

The matrix has been created, the component added in it but they are not working with the complexes

susiehgt committed:
Update complex-list.component.html

Corrections of the code

susiehgt committed:
Adding of some checkings

susiehgt committed:
Adding of Hack method

jmedinaebi commented on src/app/complex/complex-results/complex-list/complex-list.component.html:
```diff
@@ -1,3 +1,34 @@
+      </th>
+    </tr>
+    </thead>
+    <tbody *ngFor="let complex of complexSearch.elements">
```
We don't need the for loop in here

jmedinaebi commented on src/app/complex/complex-results/complex-list/complex-list.component.html:
```diff
@@ -1,3 +1,34 @@
+    </tr>
+    </thead>
+    <tbody *ngFor="let complex of complexSearch.elements">
+    <tr *ngFor="let component of complex.componentIds">
```
This for should be with the new `componentIds` variable

jmedinaebi commented on src/app/complex/complex-results/complex-list/complex-list.component.html:
```diff
@@ -1,3 +1,34 @@
+    <tr *ngFor="let component of complex.componentIds">
+      <td>{{component}}</td>
+      <td *ngFor="let complex of complexSearch.elements">
+        <ng-container *ngFor="let component of complex.componentIds">
+          x
+        </ng-container>
```
We need to replace this with a call to the new method that checks if a component in
```
<div *ngIf="doesComplexContainComponent(complex, component)">X</div>
```

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -13,6 +14,7 @@ import {AnalyticsService} from '../../shared/google-analytics/service/analytics.
   styleUrls: ['./complex-results.component.css']
 })
 export class ComplexResultsComponent implements OnInit, AfterViewInit {
+  complexResultsComponent: ComplexResultsComponent;
```
This is not needed

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -21,10 +23,14 @@ export class ComplexResultsComponent implements OnInit, AfterViewInit {
   private _spicesFilter: string[];
   private _bioRoleFilter: string[];
   private _interactorTypeFilter: string[];
+  private _allComponentsInComplexSearch: string[] = [];
```
The type should be `Set<string>`

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -42,21 +48,29 @@ export class ComplexResultsComponent implements OnInit, AfterViewInit {
+    // Calling new HACKY method to get also the components of each complex
+    this.complexPortalService.findComplexWithComponents(this.query, this.spicesFilter, this.bioRoleFilter,
       this.interactorTypeFilter, this.currentPageIndex, this.pageSize).subscribe(complexSearch => {
       this.complexSearch = complexSearch;
```
Here we need to call `this.allComponentsInComplexSearch = new Set();` to empty the set before setting all the components in it.

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -96,6 +110,10 @@ export class ComplexResultsComponent implements OnInit, AfterViewInit {
     return this._spicesFilter.length + this._interactorTypeFilter.length + this._bioRoleFilter.length;
   }
 
+  private doesComplexHaveComponent (complex, component): boolean{
```
This method needs to be moved to `ComplexListComponent`, and we need to do `complex.componentIds.includes(component)`

jmedinaebi commented on src/app/complex/complex-results/complex-list/complex-list.component.html:
```diff
@@ -1,3 +1,34 @@
+<div class="Complex-navigator">
```
In the ts file of this component, we need a new input `@Input() componentIds: Set<string>;`.

Then we can do here `*ngFor="let component of componentIds`, as we need the variable in the same component as where we use it.

Then in the HTML of ComplexResultsComponent, where we use `cp-complex-list`, we need to pass the list of component ids for the new input.
```
      <cp-complex-list [complexSearch]="complexSearch"
                       [componentIds]="allComponentsInComplexSearch"></cp-complex-list>
```

susiehgt committed:
Update on the progression of the complex portal (adding of the data in the matrix)

susiehgt committed:
Update on the progression of the complex portal (adding of the data in the matrix)

susiehgt committed:
Update

The last version was not updated correctly

jmedinaebi committed:
Add links to components

jmedinaebi committed:
Only display external link if there is one

susiehgt committed:
Adding of the subcomplexes components, not finished (the table is not yet adapted and the stoichiometry has to be added)

susiehgt committed:
Collapsible menu and stoichiometry of subcomplexes added but viewing not optimized , modification of the headers and how complexes' names are displayed

susiehgt committed:
-displaying of the collapsible menus fixed.
-stoichiometry of subcomplexes has to be fixed (display 1 when element here but not the real stoichiometry)

susiehgt committed:
-stoichiometry of the subcomplexes fixed
-displaying of the name and function of the interactors
-minor fixings on the table display
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
-fixing of the view button
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
- fixing the stoichiometry: now the main complex column also displays the stoichiometry of the interactors contained in the subcomplexes
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
- fixing the display of the subcomplexes, not it only shows the one selected
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Creation of complex-navigator directory
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Creation of complex-navigator directory
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Creation of complex-navigator directory
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Complex-navigator functional (except for see notice)
Display fixed

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Fixing of some elements of the display (icons not showing properly, size of the cells being different)

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Modification of the display of the stoichiometry
Fixing of the magnifying glass button
Changing of the display of the interactors' names
Correction of some coding elements

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

jmedinaebi committed:
Remove extra calls to fetch components, now returned from BE

susiehgt committed:
Adding of the connecting lines (not finished)
Changes on the display

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Merge branch 'complex-navigator' into complex-navigator-backend-changes

susiehgt committed:
Merge pull request #208 from Complex-Portal/complex-navigator-backend-changes

Remove extra calls to fetch components, now returned from BE

jmedinaebi committed:
Rename components as interactors

jmedinaebi committed:
Make table taller

jmedinaebi committed:
Refactor table interactor column to hide duplicated components

jmedinaebi committed:
Revert "Refactor table interactor column to hide duplicated components"

This reverts commit f531c545455d96890aab46507a4efaeaaf531df9.

jmedinaebi committed:
Rename components as interactors

jmedinaebi committed:
Simplify use of ngIfs with ng-container

jmedinaebi committed:
Create EnrichedInteractor class to store metadata about interactors displayed

jmedinaebi committed:
Refactor methods to get stochiometry

jmedinaebi committed:
Hide interactors already displayed in expanded section to avoid duplicated rows

jmedinaebi committed:
Fetch interactors in subcomplex from backend if needed

susiehgt committed:
Merge pull request #209 from Complex-Portal/complex-navigator-backend-changes

Complex navigator backend changes

susiehgt committed:
Adding of the connection lines between the circles

(missing participants not close to each other)

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Adding of the connection lines between the circles

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

jmedinaebi committed:
Use ngClass to draw the lines for the interactors

jmedinaebi committed:
Remove unused parameter

susiehgt committed:
Merge pull request #210 from Complex-Portal/complex-navigator-use-ng-class

Complex navigator use ng class

susiehgt committed:
Aesthetic corrections, alignement of the vertical lines

jmedinaebi committed:
Pre-calculate start and end indices

jmedinaebi committed:
Refactor to make logic more readable

jmedinaebi committed:
Refactor for readability

jmedinaebi committed:
Remove unused import

susiehgt committed:
Aesthetic corrections, alignement of the vertical lines

jmedinaebi committed:
Merge branch 'complex-navigator' into complex-navigator-use-ng-class

jmedinaebi committed:
Small change in CSS

susiehgt committed:
Adding of some icons, modification of the display of the complexes

susiehgt committed:
Merge pull request #211 from Complex-Portal/complex-navigator-use-ng-class

Pre-calculate start and end indices

jmedinaebi committed:
Fix wrong FTP url

jmedinaebi committed:
New npm command to publish github pages with test env

susiehgt committed:
Changing of the display of the complexes

susiehgt committed:
fix of the length of the complex name

jmedinaebi committed:
Fix issues with lines connecting interactors

susiehgt committed:
Merge pull request #212 from Complex-Portal/complex-navigator-fix-lines

Fix issues with lines connecting interactors

susiehgt committed:
Display of overflowing table header changed, still not quite optimised depending on screen size.
Adding of icons on the filter list

susiehgt committed:
Merge remote-tracking branch 'origin/complex-navigator' into complex-navigator

susiehgt committed:
Classification of the complexes depending on their similarity almost okay (need to check).
Classification of the interactors on the way

susiehgt committed:
Classification of interactors ok but lines not ok anymore

adding of line between complexes and interactors column
some fixings of the display (height and width of the matrix) -> Need to check for the stickiness of the header

susiehgt committed:
fixing of some display, header sticky but table expands to the right

susiehgt committed:
Sorting of the complexes depending on how much similar components they have with the biggest;
Changing of table display

susiehgt committed:
Short lines algorithm in progress

susiehgt committed:
Short length algorithm working on little search but crashing on big search

jmedinaebi committed:
Rename components as interactors

susiehgt committed:
Adding of a new object for binary array

susiehgt committed:
Merge pull request #213 from Complex-Portal/navigator-rename-components-as-interactors

Rename components as interactors

susiehgt committed:
Delete src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header-style2.component.css

useless

susiehgt committed:
BinaryList updated

susiehgt committed:
Sorting of the complexes by similarties OK; sorting of the components still have to include stoichiometry of subcomplexes

susiehgt committed:
Merge remote-tracking branch 'origin/complex-navigator' into complex-navigator

# Conflicts:
#	src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts

susiehgt committed:
Classifying interactors by appearance ok

susiehgt committed:
Adding of the ranges for interactors types and organism;
Button for choosing the types of listing of the interactors not functional, but functions work separatly

susiehgt committed:
Fixing of the sticky header and the clickability of the complexes header

susiehgt committed:
Beginning of adding lines depending on interactors listing

susiehgt committed:
Interactors classification button + sidebars OK

susiehgt committed:
Fixing of the size of the
table and scrolling issues caused by it;
Fixing of typos;
Interactors Occurence sorting button to check (scrolling refreshing makes the whole table lagging by playing the sorting function);
Sorting buttons to check (have to click multiple time to work)

jmedinaebi commented on angular.json:
```diff
@@ -26,6 +26,7 @@
               "src/favicon.ico"
             ],
             "styles": [
+              "./node_modules/@angular/material/prebuilt-themes/deeppurple-amber.css",
```
What is this style and why do we need it?

susiehgt replied:
I have no idea, I've deleted it and nothing seems to change

jmedinaebi commented on karma.conf.js:
```diff
@@ -1,5 +1,6 @@
 // Karma configuration file, see link for more information
 // https://karma-runner.github.io/0.13/config/configuration-file.html
+// test
```
I'm guessing all the changes in this file can be reverted.

jmedinaebi commented on package.json:
```diff
@@ -27,40 +28,44 @@
     "expose-loader": "^0.7.5",
     "jquery": "^3.4.1",
     "litemol": "https://github.com/dsehnal/LiteMol/tarball/master",
+    "ml-kmeans": "^6.0.0",
```
Some of these new dependencies seem related to clustering. Are we using all of them? Or did you add some for some testing but ended up not using them? We should remove the new dependencies we're not using.

susiehgt replied:
I have added them during the time I was looking for an algorithm which would do the clustering, I will check what is useless so I will remove them

jmedinaebi commented on src/app/complex/complex-results/complex-list/complex-list.component.html:
```diff
@@ -1,14 +1,14 @@
 <ul class="no-bullet">
   <li *ngFor="let complex of complexSearch.elements">
-    <h3><a [routerLink]="['/complex', complex.complexAC]">{{complex.complexName}}</a></h3>
+    <h3><a [routerLink]="['/complex', complex.complexAC]">{{ complex.complexName }}</a></h3>
```
These type of changes make sense on files we are changing, like `complex-filter.component.html`, but these changes in this file or in `complex-filter.component.spec.ts` just make the review more difficult, as there are mor files to review and it's harder to check which files have proper changes vs files with just spacing changes.

susiehgt replied:
Okay, I will pay attention because it's the autocorrecting from webstorm which is reformatting some parts

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator.component.ts:
```diff
@@ -0,0 +1,41 @@
+})
+export class ComplexNavigatorComponent implements OnInit {
+  @Input() complexSearch: ComplexSearchResult;
+  _interactorsSorting: string;
```
We don't need to set and get method for `_interactorsSorting` and `_interactors`. We can simply do
```
@Input() interactorsSorting: string;
@Input() interactors: Set<Interactor>;
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.css:
```diff
@@ -0,0 +1,135 @@
+/* Header */
+
+.CN-table {
+  transform: rotateX(180deg); /* On reinverse le contenu du header pour que le contenu soit dans le bon sens */
```
Could you translate the comment?

susiehgt replied:
We rotate the header's content to make it it the right sense (when removes, it is upside down)

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.css:
```diff
@@ -0,0 +1,135 @@
+}
+
+
+/*.line {*/
```
Delete

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.html:
```diff
@@ -0,0 +1,49 @@
+                   target="_blank">
+                    <div>
+            <span title="{{ complex.complexName }} - {{ complex.complexAC}}">
+<!--            <div class="line"></div>-->
```
I guess we can delete this commented line.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.ts:
```diff
@@ -0,0 +1,148 @@
+      let totalLength = complex.interactors.length;
+      for (const complexInteractorChecked of complex.interactors) {
+        if (complexInteractorChecked.interactorType === 'stable complex') {
+          // tslint:disable-next-line:no-shadowed-variable
```
We should avoid these type of `tslint` comments. Here it is complaining because `complex` was defined on line 30 and now again inside the `find` method. We can simply rename it inside the method to something like
```
const subComplex: Element = searchResult.find(c => c.complexAC === complexInteractorChecked.identifier);
```
And then we don't need the comment.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.ts:
```diff
@@ -0,0 +1,148 @@
+      }
+    }
+
+    const bigComplex: any = biggestComplex[0]; // access to the complex
```
We know the type is not any, right?

susiehgt replied:
I have a problem with this one, the biggestComplex[0] is an Element, so logically I assign the type Element to bigComplex, but it raise an error saying that it should be a number

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.ts:
```diff
@@ -0,0 +1,148 @@
+        }
+        if (biggestComplexInteractor.interactorType === 'stable complex') {
+          // tslint:disable-next-line:max-line-length
+          const subComplex: Element = searchResult.find(complex => complex.complexAC === biggestComplexInteractor.identifier);
```
Same as above.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.ts:
```diff
@@ -0,0 +1,148 @@
+        if (biggestComplexInteractor.interactorType === 'stable complex') {
+          // tslint:disable-next-line:max-line-length
+          const subComplex: Element = searchResult.find(complex => complex.complexAC === biggestComplexInteractor.identifier);
+          if (comparedComplex.complexAC === bigComplex.complexAC) {
```
I'm not sure about this logic, I'll take a deeper look at it. I'm not sure if it's entirely correct and if so, maybe it can be optimised.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.ts:
```diff
@@ -0,0 +1,148 @@
+      complexesWithSimilarities.set(comparedComplex, similarities);
+    }
+    // sort complexes depending on their similarities with the biggest
+    this.complexSearch.elements.sort((a, b) => complexesWithSimilarities.get(b)! - complexesWithSimilarities.get(a)!);
```
What does the `!` do here?

susiehgt replied:
It is to make sure the value it fetches is not null

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.css:
```diff
@@ -0,0 +1,119 @@
+  content: '';
+  position: absolute;
+  inset: -2px -1px -2px -1px;
+  /*background: #0e6f7612;*/
```
Delete commented out lines

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.ts:
```diff
@@ -0,0 +1,148 @@
+
+  ngOnInit(): void {
+    this.classifyComplexesSize();
+    this.classifyComplexesSimilarities(this.complexSearch);
```
What does the second method do? I understand `classifyComplexesSize`, but not `classifyComplexesSimilarities`.

susiehgt replied:
ClassifyComplexSimilarities is used to cluster the complexes by number of similarities, it puts the similar complexes next to each other in the list which is used to sort the complexes

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.html:
```diff
@@ -0,0 +1,128 @@
+<div class="Complex-navigator">
```
This component has gotten too big and we should split it into multiple ones. Don't worry about it for now, as we may need to use events when we split it on multiple components. I'll take a look at splitting this and then we can discuss it next week.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:
```diff
@@ -0,0 +1,41 @@
+})
+export class TableStructureComponent implements OnInit {
+  @Input() complexSearch: ComplexSearchResult;
+  _interactors: Set<Interactor>;
```
Same as above, we don't need the `@Input` methods.

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.html:
```diff
@@ -1,41 +1,77 @@
+          <cp-complex-list [complexSearch]="complexSearch">
+          </cp-complex-list>
+        </div>
+        <div class="Complex-navigator" *ngIf="DisplayType">
```
The same way we have a component `cp-complex-list`, we should have a component for the complex navigator with the sorting and the table all in that component, and here we simply do
```
<cp-complex-navigator [complexSearch]="complexSearch"
                                  [interactors]="allInteractorsInComplexSearch"
                                  [interactorsSorting]="interactorsSorting">
            </cp-complex-navigator>
```
I guess one issue with that is that we have the toggle display button with the sorting buttons, but I think they should remain separate. We should have one (or even 2, depending on how we do it) toggle button above to change the view, and then, after that, the sorting buttons.

jmedinaebi replied:
Then we can move all the logic and html about the sorting to the `cp-complex-navigator` component.

jmedinaebi replied:
As mentioned above, we can move the sorting buttons to `cp-complex-navigator`, and have the toggle button above all of this.

I have made these changes locally by doing:

- In `complex-results.component.ts`
```
  LIST_VIEW = 'LIST_VIEW';
  COMPLEX_NAVIGATOR_VIEW = 'COMPLEX_NAVIGATOR_VIEW';
  ...
  DisplayType: string = this.LIST_VIEW;
  ...
 // methods to replace toggleDisplayType()
  setListView() {
    this.DisplayType = this.LIST_VIEW;
  }

  setComplexNavigatorView() {
    this.DisplayType = this.COMPLEX_NAVIGATOR_VIEW;
  }
```

- In `complex-results.component.html`, at the very top
```
  <ng-container *ngIf="complexSearch;else loadingSpinner">
    <div class="columns medium-12" *ngIf="complexSearch.totalNumberOfResults !== 0">
      <h2 class="padding-left-large">Total number of results: {{ complexSearch.totalNumberOfResults }} </h2>

      <div class="row">
        <div class="displayButton columns medium-8" style="text-align: center">
          <input type="submit" name="submit" value="List View" class="button view-button"
                 (click)="setListView()" [ngClass]="{'selected': DisplayType === LIST_VIEW}">
          <input type="submit" name="submit" value="Complex Navigator View" class="button view-button"
                 (click)="setComplexNavigatorView()" [ngClass]="{'selected': DisplayType === COMPLEX_NAVIGATOR_VIEW}">
        </div>
      </div>
     ...
```
And then delete the following chunks
```
          <div class="displayButton" style="text-align: center">
            <input type="submit" name="submit" value="Complex navigator" class="button"
                   (click)="toggleDisplayType()">
          </div>
```
```            <div class="displayButton" style="text-align: center">
              <input type="submit" name="submit" value="View list" class="button"
                     (click)="toggleDisplayType()">
            </div>
```
And update the ngIf to be
```
<div class="listOfResults" *ngIf="DisplayType===LIST_VIEW">
```
```
<div class="Complex-navigator" *ngIf="DisplayType===COMPLEX_NAVIGATOR_VIEW">
```

- And finally in the `complex-results.component.css` file, add
```
.displayButton {
  float: right;
}

.view-button {
  float: left;
  width: 225px;
  margin-right: 5px;
  background-color: white;
  color: #555;
  border: 1px dotted #007c82;
}

.view-button:hover {
  color: #007c82;
  background-color: rgb(234, 234, 234);
  border: transparent;
  border-bottom: 1px dotted #007c82;
}

.selected,
.selected:hover {
  background-color: #007c82;
  color: white;
  opacity: 100%;
}
```

jmedinaebi replied:
Then it would look like this 

https://github.com/Complex-Portal/complex-portal-view/assets/114929382/2860832b-360e-48a5-83b3-55bc4c0e0470

jmedinaebi commented on src/environments/environment.test.ts:
```diff
@@ -7,7 +7,7 @@ const EBI_BASE_URL = 'https://www.ebi.ac.uk/';
 const LICENSE_URL = 'https://creativecommons.org/licenses/by/4.0/';
 const REACTOME_BASE_URL = 'https://reactome.org';
-const INTACT_FTP_BASE_URL = 'ftp://ftp.ebi.ac.uk/pub/databases/intact/';
+const INTACT_FTP_BASE_URL = 'http://ftp.ebi.ac.uk/pub/databases/intact/';
```
Revert this change.

jmedinaebi commented on src/index.html:
```diff
@@ -69,6 +69,9 @@
   -->
 
   <!-- end CSS-->
+  <link rel="preconnect" href="https://fonts.gstatic.com">
```
What are these changes?

jmedinaebi commented on src/styles.css:
```diff
@@ -12,3 +12,6 @@
 .button.secondary-background:hover { background: #666; }
 .button.medium-gray-background:hover { background: #999; }
+
+html, body { height: 100%; }
```
What are these changes?

susiehgt replied:
I guess this is from when I was trying to make the header sticky, I have removed it now because it does not change anything

jmedinaebi reviewed (commented):
I had a look at everything but the `TableInteractorColumnComponent` component. I know that's where most of the logic is, but I think we need to split that into multiple components and also review it. I'll take a proper look at that component locally and I'll let you know my comments after that.

jmedinaebi commented on src/app/app.component.css:
```diff
@@ -24,5 +24,23 @@
+  position: sticky;
+  top: 0;
+  background: #007c82;
+  margin-left: var(--global-margin);
```
There is a problem with this. Normally, this left margin is a negative value, which makes the container to move all the way to the left, but if the window is too small, it ends up being a positive value and pushes the container to the right, creating a weird empty space on the left.

This can be fixed by using `min`:
```
margin-left: min(var(--global-margin), 0px);
```

Check the video in https://drive.google.com/file/d/1DuU0VFE-9pbvfuXSmgO0aQEM3f4MnJIn/view?usp=drive_link to see how it currently behaves when you resize the window. This does not happen after the change suggested. I tried to attach the video here but I couldn't, as it is too big.

jmedinaebi replied:
Already done

susiehgt committed:
Ranges of interactor type V2, not working when expanded

susiehgt committed:
Ranges of interactor type V2, working but some bugs on the expanded complex

susiehgt committed:
Ranges of interactor type V2, working but some bugs on the first expanded complex

susiehgt committed:
Modification of the classification divs

susiehgt committed:
Fixing of some classification bugs

susiehgt committed:
Fixing of interactors displaying

jmedinaebi commented on src/app/complex/complex-results/complex-filter/complex-filter.component.ts:
```diff
@@ -133,4 +134,30 @@ export class ComplexFilterComponent implements OnInit {
     this._interactorTypeFilter = value;
   }
+
+  public interactorTypeIcon(facet): string {
```
As we did with the `complex-navigator-utils.ts` file, we should have a file `complex-portal-utils.ts`, probably located at `src/app/complex/` where we can put the `interactorTypeIcon` function. It is used here and then in the complex navigator, so let's extract it to a common place.

susiehgt committed:
code reorganisation and optimisation, fixing of display elements

susiehgt committed:
Creation of complex portal utils, changing of icons function.
Start to implement the adjusting of the interactors header to the size of the cells when sorting is on §§§§

susiehgt committed:
Creation of complex portal utils, changing of icons function.
Start to implement the adjusting of the interactors header to the size of the cells when sorting is on §§§§

jmedinaebi commented on angular.json:
```diff
@@ -108,6 +108,7 @@
               "node_modules/complexviewer/dist/complexviewer.js"
             ],
             "styles": [
+              "./node_modules/@angular/material/prebuilt-themes/deeppurple-amber.css",
```
Is this needed?

susiehgt replied:
I thought I've removed it

jmedinaebi commented on package.json:
```diff
@@ -42,25 +45,25 @@
     "@types/foundation": "~5.5.30",
     "@types/jasmine": "~3.6.0",
-    "@types/jquery": "~3.3.36",
+    "@types/jquery": "^3.3.38",
```
Any reason to change the jquery version?

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.ts:
```diff
@@ -0,0 +1,152 @@
+  styleUrls: ['./table-header.component.css'],
+})
+
+export class TableHeaderComponent implements OnInit {
```
I'm not sure about the methods in this class, I'll take a deeper look at them and then we can discuss it.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.ts:
```diff
@@ -0,0 +1,152 @@
+    });
+  }
+
+  getInteractorSorting() {
```
This should be renamed to indicate it returns a boolean, something like `isInteractorSortingSet`.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts:
```diff
@@ -0,0 +1,391 @@
+    }
+  }
+
+  get rangesOfInteractorType(): number[] {
```
Let's rename `_rangesOfInteractorsType` as `rangesOfInteractorsType` and get rid of this method. Same for the method below, and for `enrichedInteractors` and `enrichedComplexes`.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts:
```diff
@@ -0,0 +1,391 @@
+    this.interactorOrganism();
+  }
+
+  stoichiometryOfInteractors(complex: Element, interactorId: string): string {
```
Tis method exists in multiple components, let's move it to `complex-navigator-utils`.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-main-interactor/table-main-interactor.component.ts:
```diff
@@ -0,0 +1,196 @@
+    return null;
+  }
+
+  stoichiometryOfInteractors(complex: Element, interactorId: string): string {
```
Duplicated method mentioned before.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-subcomponent-interactor/table-subcomponent-interactor.component.ts:
```diff
@@ -0,0 +1,153 @@
+    return null;
+  }
+
+  stoichiometryOfInteractors(complex: Element, interactorId: string): string {
```
Duplicated method mentioned before.

jmedinaebi commented on src/app/complex/complex-results/complex-filter/complex-filter.component.ts:
```diff
@@ -133,4 +137,8 @@ export class ComplexFilterComponent implements OnInit {
     this._interactorTypeFilter = value;
   }
+
+  public facetTypeIcon(facet): string {
```
This works, but we can also do:
```
interactorTypeIcon = interactorTypeIcon;
```

And then in the html
```
<i class="{{ interactorTypeIcon(facet.name) }}"
                 style="font-size: large;"></i>
```

Either way it works.

jmedinaebi commented on src/app/complex/complex-results/complex-filter/complex-filter.component.ts:
```diff
@@ -14,9 +16,11 @@ export class ComplexFilterComponent implements OnInit {
   private _bioRoleFilter: string[];
   private _interactorTypeFilter: string[];
 
+  @Input() icon: string;
```
Is this input needed?

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts:
```diff
@@ -0,0 +1,391 @@
+
+  _enrichedInteractors: EnrichedInteractor[];
+  _enrichedComplexes: EnrichedComplex[];
+  _rangesOfInteractorsType: number[];
```
We actually don't need `_rangesOfInteractorsType` and `_rangesOfInteractorsOrganism`. We could simple have `ranges: number[]`, and then in lines 363 and 389 do `this.ranges = ranges;`

ranges gets sets with the right values based on the sorting selected, and then we don't need the `ranges()` function either.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts:
```diff
@@ -0,0 +1,391 @@
+        for (const complexesInteractors of complex.interactors) {
+          if (oneInteractor.interactor.identifier === complexesInteractors.identifier) {
+            // tslint:disable-next-line:radix
+            if (isNaN(parseInt(this.stoichiometryOfInteractors(complex, oneInteractor.interactor.identifier)))) {
+              oneInteractor.timesAppearing = oneInteractor.timesAppearing;
+            } else {
+              oneInteractor.timesAppearing += parseInt(this.stoichiometryOfInteractors(complex, oneInteractor.interactor.identifier), 10);
+            }
```
```
if (!isNaN(parseInt(this.stoichiometryOfInteractors(complex, oneInteractor.interactor.identifier)))) {
              oneInteractor.timesAppearing += parseInt(this.stoichiometryOfInteractors(complex, oneInteractor.interactor.identifier), 10);
            }
```

jmedinaebi replied:
Actually, instead of calling `this.stoichiometryOfInteractors` twice, we could do:
```
const stoichiometryValue = parseInt(this.stoichiometryOfInteractors(complex, oneInteractor.interactor.identifier), 10);
if (!isNaN(stoichiometryValue)) {
    oneInteractor.timesAppearing += stoichiometryValue;
}
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts:
```diff
@@ -0,0 +1,391 @@
+            }
+          }
+        }
+        if (oneInteractor.isSubComplex) {
```
This line is what's causing the table to be empty sometimes. Change it to
```
if (oneInteractor.isSubComplex && !!oneInteractor.subComponents) {
```

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.html:
```diff
@@ -1,41 +1,65 @@
+          <cp-complex-list [complexSearch]="complexSearch">
+          </cp-complex-list>
+        </div>
+        <div class="Complex-navigator" *ngIf="DisplayType===COMPLEX_NAVIGATOR_VIEW">
```
I would change this to
```
            <cp-complex-navigator class="Complex-navigator" *ngIf="DisplayType===COMPLEX_NAVIGATOR_VIEW" 
                                  [complexSearch]="complexSearch"
                                  [interactors]="allInteractorsInComplexSearch">
            </cp-complex-navigator>
```

And move `interactorsSorting` variable to the complex navigator component and update the inputs. Then in the complex navigator html do
```
<div class="ComplexNavigator">
  <cp-complex-navigator-sorting
            [(interactorsSorting)]="interactorsSorting">
  </cp-complex-navigator-sorting>
  <div class="Matrix">
    <cp-table-structure [complexSearch]="complexSearch"
                        [interactors]="interactors"
                        [interactorsSorting]="interactorsSorting">
    </cp-table-structure>
  </div>
</div>
```

jmedinaebi reviewed (commented):
Try to get rid of the `tslint:disable...` comments.

susiehgt committed:
Removing of the tslint comments

susiehgt committed:
Removing of useless code, fixing of some duplications

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.ts:
```diff
@@ -0,0 +1,158 @@
+    });
+  }
+
+  isInteractorSortingSet() {
```
This can be rewritten as
```
  isInteractorSortingSet() {
    return this.interactorsSorting === 'Type' || this.interactorsSorting === 'Organism';
  }
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-external-link/table-interactor-external-link.component.ts:
```diff
@@ -0,0 +1,27 @@
+import {Component, Input, OnInit} from '@angular/core';
+import {Interactor} from '../../../../../shared/model/complex-results/interactor.model';
```
Some imports here are unused.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-subcomponent-interactor/table-subcomponent-interactor.component.ts:
```diff
@@ -0,0 +1,153 @@
+  constructor() {
+  }
+
+  ngOnInit(): void {
```
There's a few empty ngOnInit methods. We could get rid of them and the `implements OnInit` on the class name and it should be fine.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts:
```diff
@@ -0,0 +1,361 @@
+    for (const complex of this.complexSearch.elements) {
+      const organismName = complex.organismName;
+      for (const complexInteractor of complex.interactors) {
+        // tslint:disable-next-line:max-line-length
```
tslint:disable comment, please fix and delete.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:
```diff
@@ -0,0 +1,23 @@
+})
+export class TableStructureComponent implements OnInit {
+  @Input() complexSearch: ComplexSearchResult;
+  @Input() interactors: Set<Interactor>;
```
This should be of type `Interactor[]` instead of set. It's complaining about it.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator.component.ts:
```diff
@@ -0,0 +1,22 @@
+export class ComplexNavigatorComponent implements OnInit {
+  @Input() complexSearch: ComplexSearchResult;
+  @Input() interactorsSorting: string;
+  @Input() interactors: Set<Interactor>;
```
This should be of type Interactor[] instead of set. It's complaining about it.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator.component.ts:
```diff
@@ -0,0 +1,22 @@
+})
+export class ComplexNavigatorComponent implements OnInit {
+  @Input() complexSearch: ComplexSearchResult;
+  @Input() interactorsSorting: string;
```
`interactorsSorting` does not need to be an input. We could simply do `interactorsSorting = 'Occurrence';` here and then remove the input in the html.

jmedinaebi replied:
In the complex-results.component.html, we then do:
```
        <cp-complex-navigator class="Complex-navigator" *ngIf="DisplayType===COMPLEX_NAVIGATOR_VIEW"
                              [complexSearch]="complexSearch"
                              [interactors]="allInteractorsInComplexSearch">
        </cp-complex-navigator>
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.html:
```diff
@@ -0,0 +1,87 @@
+              [ngClass]="{'firstExpandedRow': j === 0, 'lastExpandedRow': j === interactor.subComponents.length - 1}">
+            <td [attr.rowspan]="interactor.subComponents.length" class="interactorSeparation"
+                title="{{interactor.interactor.name}}'s interactors"
+                *ngIf="(interactorsSorting === 'Type' && j===0) || (interactorsSorting === 'Organism' && j===0)">
```
This if can be simplified.
```
*ngIf="j === 0 && (interactorsSorting === 'Type' || interactorsSorting === 'Organism')">
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts:
```diff
@@ -0,0 +1,361 @@
+      return of(foundComplex.interactors);
+    } else {
+      // Actually call the back-end to fetch these
+      return this.complexPortalService.getComplexAc(interactor.interactor.identifier)
```
I have deployed the changes in the back-end. What we need to do now is,

In complex-portal.service.ts we create a new method
```
  getSimplifiedComplex(complexAc: string): Observable<Element> {
    const url = `${baseURL}/complex-simplified/${complexAc}`;
    // const url = `${baseURL}/complex-simplified/${complexAc}`;
    return this.http.get(url).pipe(
      catchError(this.handleError));
  }
```

And then here we replace this call (lines 145-152) with:
```
      return this.complexPortalService.getSimplifiedComplex(interactor.interactor.identifier)
        .pipe(map(complex => complex?.interactors));
```

jmedinaebi reviewed (commented):
A few comments with small changes and also the change to call the new back-end endpoint.

susiehgt committed:
Calling from backend changed, removing of useless imports and ngOnInit

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:
```diff
@@ -0,0 +1,18 @@
+  @Input() interactors: Interactor[];
+  @Input() interactorsSorting: string;
+
+  constructor(private router: Router) {
```
`router` is not used in this component, so we can rid of this constructor.

susiehgt committed:
Fixing of displaying bugs while using filters and/or changing display type,
adding of the view in the IRL,
adding of the icons for the species,
adding of pagination tools (first/last page)

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator.component.ts:
```diff
@@ -0,0 +1,14 @@
+export class ComplexNavigatorComponent {
+  @Input() complexSearch: ComplexSearchResult;
+  @Input() interactors: Interactor[];
+  interactorsSorting: string = 'Occurrence';
```
When a variable is set and the type is clear, like a string or a number, typescript doesn't need the type and complains about it actually.

Here, because `'Occurrence'` is clearly a string, we don't need `:string`, we can simply do.
```
interactorsSorting = 'Occurrence';
```

It's the same with numbers
```
const a: number = 1; // typescript does not like it
const b = 1; // typescript is happy
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-name/table-interactor-name.component.html:
```diff
@@ -0,0 +1,15 @@
+  <!-- Adding of the icons to access details of the interactor -->
+  <a [routerLink]="['/complex/search']"
+     [queryParams]="{query: interactorId, page: 1}"
+     [fragment]="'NAVIGATOR_VIEW'"
```
Instead of `NAVIGATOR_VIEW` we need to use `view_complex_navigator`, which is the value `complex-results.component` understand.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-name/table-interactor-name.component.ts:
```diff
@@ -0,0 +1,29 @@
+  @Input() interactorName: string;
+  @Input() interactorType: string;
+  @Input() interactorId: string;
+  @Input() interactorTypeIcon: string;
```
`interactorTypeIcon` and `interactorOrganismIcon` are not inputs.

```
  @Input() interactorName: string;
  @Input() interactorType: string;
  @Input() interactorId: string;
  @Input() interactorOrganism: string;

  interactorTypeIcon: string;
  interactorOrganismIcon: string;
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-name/table-interactor-name.component.ts:
```diff
@@ -0,0 +1,29 @@
+  }
+
+  formatInteractorOrganism() {
+    let from = this.interactorOrganism.indexOf(';');
```
`from` does not get updated, change `let` to `const`.

jmedinaebi commented on src/app/complex/complex-results/complex-filter/complex-filter.component.ts:
```diff
@@ -1,6 +1,8 @@
 import {Component, EventEmitter, Input, OnInit, Output} from '@angular/core';
 import {Facets} from '../../shared/model/complex-results/facets.model';
 import {AnalyticsService} from '../../../shared/google-analytics/service/analytics.service';
+import {Interactor} from '../../shared/model/complex-results/interactor.model';
```
Unused import

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.ts:
```diff
@@ -0,0 +1,18 @@
+})
+
+export class TableHeaderComponent {
+  // @Input() complexSearch: ComplexSearchResult;
```
Delete the commented line

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts:
```diff
@@ -0,0 +1,345 @@
+        const match = this.enrichedInteractors.find(
+          enrichedInteractor => enrichedInteractor.interactor.identifier === complexInteractor.identifier
+        );
+        match.organismName = organismName;
```
```
        if (!!match) {
          match.organismName = organismName;
        }
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts:
```diff
@@ -0,0 +1,345 @@
+              enrichedInteractor => enrichedInteractor.interactor.identifier === oneSubInteractor.identifier
+            );
+            if (!!oneEnrichedInteractor) {
+              oneEnrichedInteractor.timesAppearing = parseInt(formatStoichiometryValues(oneSubInteractor.stochiometry), 10);
```
Just to be safe that the stoichiometry is actually a number
```
              const stoichiometryValue = parseInt(formatStoichiometryValues(oneSubInteractor.stochiometry), 10);
              if (!isNaN(stoichiometryValue)) {
                oneEnrichedInteractor.timesAppearing += stoichiometryValue;
              }
```

jmedinaebi commented on src/app/complex/complex-results/complex-paginator/complex-paginator.component.html:
```diff
@@ -1,12 +1,15 @@
+  <h5 class="text-center">{{ currentPageIndex }} of {{ lastPageIndex }}</h5>
   <ul class="pagination text-center" role="navigation" aria-label="Pagination">
+    <li class="arrow unavailable" *ngIf="currentPageIndex != 1" aria-disabled="true"><a
+      (click)="getFirstPage()">&laquo; First</a></li>
```
Currently we are using `&laquo;` for first and previous page and `&raquo` for next and last page, meaning we are using the same icon for first and previous, and the same icon form next and last.

We can change it to:
- First: `&Lang;`
- Previous: `&lang;`
- Next: `&rang;`
- Last: `&Rang;`

That way we have different icons for each button.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/complex-navigator-utils.ts:
```diff
@@ -0,0 +1,47 @@
+  return ' '; // sometimes we don't have the stoichiometry value
+}
+
+export function fetchValuesFromStoichiometry(stoichiometry: string) {
```
Right now this function is used in multiple places with similar code. We can make some changes to reduce code duplication.

First, we have the method `formatStoichiometryValues` in this same class.
Then, the method `addedStoichiometryValues` in table-main-interactor.component.ts
These 2 methods are quite similar and we can reduce code duplication by doing.

In this class, add a new method
```
export function parseStoichiometryValues(stoichiometry: string): [number, number] {
  if (!!stoichiometry) {
    const matchedStoichiometry = fetchValuesFromStoichiometry(stoichiometry);
    if (!!matchedStoichiometry) {
      const minValue = parseInt(matchedStoichiometry[1], 10);
      const maxValue = parseInt(matchedStoichiometry[2], 10);
      if (!isNaN(minValue) && !isNaN(maxValue)) {
        return [minValue, maxValue];
      }
    }
  }
  return null;
}
```

And also in this class update `formatStoichiometryValues`
```
export function formatStoichiometryValues(stoichiometry: string): string {
  const parsedStoichiometry = parseStoichiometryValues(stoichiometry);
  if (!!parsedStoichiometry) {
    const minValue = parsedStoichiometry[0];
    const maxValue = parsedStoichiometry[1];
    if (minValue === maxValue) {
      return minValue.toString();
    } else {
      return `${minValue}, ${maxValue}`;
    }
  }
  return ' '; // sometimes we don't have the stoichiometry value
}
```

Finally, in table-main-interactor.component.ts, update `addedStoichiometryValues` to
```
  private addedStoichiometryValues(components: ComplexComponent[]): [number, number] {
    let minValue: number = null;
    let maxValue: number = null;
    for (const component of components) {
      const parsedStoichiometry = parseStoichiometryValues(component.stochiometry);
      if (!!parsedStoichiometry) {
        if (minValue === null) {
          minValue = 0;
        }
        if (maxValue === null) {
          maxValue = 0;
        }
        minValue += parsedStoichiometry[0];
        maxValue += parsedStoichiometry[1];
      }
    }
    if (minValue !== null && maxValue !== null) {
      return [minValue, maxValue];
    }
    return null;
  }
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/complex-navigator-utils.ts:
```diff
@@ -0,0 +1,47 @@
+}
+
+export function getStoichiometry(complex: Element, componentId: string): string {
+  const match = this.findInteractorInComplex(complex, componentId);
```
`const match = findInteractorInComplex(complex, componentId);`

The `this.` doesn't work here.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-subcomponent-interactor/table-subcomponent-interactor.component.html:
```diff
@@ -0,0 +1,12 @@
+<div [ngClass]="displayTopLineClassExpanded(complex, i, j)"></div>
```
Because of how angular works, methods call in `ngClass` and `ngIf` are called periodically even if nothing changes, which is a bit inefficient. One way to work around this is to call these methods in `ngOnChanges`, set the values to variables and then use the variables in the HTML.

For example, for this component, the ts file would be:
```
...
export class TableSubcomponentInteractorComponent implements OnChanges {
  @Input() complex: EnrichedComplex;
  @Input() i: number;
  @Input() j: number;
  @Input() enrichedInteractors: EnrichedInteractor[];

  displayTopLineClass: string;
  displayBottomLineClass: string;
  interactorInComplex = false;
  interactorStoichiometryText: string;
  interactorStoichiometryValue: string;
  interactorInSubComplex = false;
  subComplexInteractorStoichiometryText: string;
  subComplexInteractorStoichiometryValue: string;

  // findInteractorInComplex = findInteractorInComplex;
  // getStoichiometry = getStoichiometry;
  // stoichiometryOfInteractors = stoichiometryOfInteractors;

  ngOnChanges(changes: SimpleChanges): void {
    this.displayTopLineClass = this.displayTopLineClassExpanded(this.complex, this.i, this.j);
    this.interactorInComplex = !!findInteractorInComplex(this.complex.complex, this.el.identifier);
    this.interactorStoichiometryText = getStoichiometry(this.complex.complex, this.el.identifier);
    this.interactorStoichiometryValue = stoichiometryOfInteractors(this.complex.complex, this.el.identifier);
    this.interactorInSubComplex = !!this.findInteractorInExpandedSubComplex(this.interactor, this.complex.complex, this.el.identifier);
    this.subComplexInteractorStoichiometryText = this.getStoichiometryInExpandedSubComplex(this.interactor, this.el.identifier);
    this.subComplexInteractorStoichiometryValue = this.stoichiometryOfInteractorsExpandable(this.interactor, this.el.identifier);
    this.displayBottomLineClass = this.displayBottomLineClassExpanded(this.complex, this.i, this.j);
  }
...
```

Then, the HTML would be
```
<div [ngClass]="displayTopLineClass"></div>
<cp-table-interactor-stoichiometry
  *ngIf="interactorInComplex"
  [stoichiometryText]="interactorStoichiometryText"
  [stoichiometryValue]="interactorStoichiometryValue"
></cp-table-interactor-stoichiometry>
<cp-table-interactor-stoichiometry
  *ngIf="interactorInSubComplex"
  [stoichiometryText]="subComplexInteractorStoichiometryText"
  [stoichiometryValue]="subComplexInteractorStoichiometryValue"
></cp-table-interactor-stoichiometry>
<div [ngClass]="displayBottomLineClass"></div>
```

These changes mean that the methods are only called when there are changes on the inputs, and not periodically. These changes are not needed, but if you want to improve performance a bit, this type of change on different components could do that.

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.html:
```diff
@@ -1,41 +1,57 @@
+      <h2 class="padding-left-large">Total number of results: {{ complexSearch.totalNumberOfResults }} </h2>
+      <div class="row">
+        <div class="displayButton columns medium-8" style="text-align: center">
+          <input type="submit" name="submit" value="View list" class="button view-button"
```
For when we launch, we probably want the list view to be the default one, as it is the one the users are used to. What we can do is leave the list view as the default one, but highlight that we have a new view, with something like.

![Screenshot 2024-04-30 at 14 19 24](https://github.com/Complex-Portal/complex-portal-view/assets/114929382/f3ee2956-1a31-4aa3-b50e-72f46ec598ff)

![Screenshot 2024-04-30 at 14 19 29](https://github.com/Complex-Portal/complex-portal-view/assets/114929382/acb28056-e058-4861-9f40-a7ee0515aa23)

What I did to have this is, first modify the HTML to add the new text and classes
```
          <button type="submit" name="submit" class="button view-button"
                 (click)="setListView()" [ngClass]="{'selected': DisplayType === LIST_VIEW}">
            View list
          </button>
          <button type="submit" name="submit" class="button view-button"
                 (click)="setComplexNavigatorView()" [ngClass]="{'selected': DisplayType === COMPLEX_NAVIGATOR_VIEW}">
            View Complex Navigator
            <span class="new-tag" [ngClass]="{'new-tag-selected': DisplayType === COMPLEX_NAVIGATOR_VIEW}">NEW!</span>
          </button>
```

I had to change the `<input>` to `<button>` to be able to have elements inside it.

Then, add new classes to the CSS
```
.new-tag {
  color: red;
  font-weight: bold;
  font-size: x-small;
}

.new-tag-selected {
  color: yellow;
}
```

Give it a try if you think it's worth doing this.

jmedinaebi reviewed (commented):
Regarding how to display hide things, what I did was.

First, add the checkboxes in complex-navigator.component, although maybe it can be its separate component.
In the ts file:
```
...
  displayInteractorOrganism = true;
  displayInteractorType = true;
  displayInteractorIdentifier = true;

  changeOrganismFlag() {
    this.displayInteractorOrganism = !this.displayInteractorOrganism;
  }

  changeInteractorTypeFlag() {
    this.displayInteractorType = !this.displayInteractorType;
  }

  changeIdentifierFlag() {
    this.displayInteractorIdentifier = !this.displayInteractorIdentifier;
  }
...
```
In the HTML file
```
  ...
  <cp-complex-navigator-sorting
    [(interactorsSorting)]="interactorsSorting">
  </cp-complex-navigator-sorting>
  <div class="displayOptionsDiv">
    <input type="submit" name="displayOptions" value="Display options" class="button">
    <div class="displayOptions">
      <input type="checkbox" value="Organism"
             (change)="changeOrganismFlag()" [checked]="displayInteractorOrganism">Organism
      <input type="checkbox" value="Interactor Type"
             (change)="changeInteractorTypeFlag()" [checked]="displayInteractorType">Interactor Type
      <input type="checkbox" value="Identifier"
             (change)="changeIdentifierFlag()" [checked]="displayInteractorIdentifier">Identifier
    </div>
  </div>
...
```

I added a new input to `table-interactor-external-link.component`:
```
@Input() displayInteractorIdentifier: boolean;
```
WIth the HTML being now
```
<ng-container *ngIf="showExternalLink()">
  <div class="externalLinkContainer" *ngIf="displayInteractorIdentifier">
    <a href="{{identifierLink}}"
       target="_blank">{{ interactorId }}
      <i class="icon icon-generic small" data-icon="x" title="More information"></i>
    </a>
  </div>
  <a *ngIf="!displayInteractorIdentifier" href="{{identifierLink}}"
     target="_blank">
    <i class="icon icon-generic small" data-icon="x" title="More information"></i>
  </a>
</ng-container>

```

Then, I moved `cp-table-interactor-external-link` to `table-interactor-name.component`, and also added some ngIf for other stuff.
```
<div>
  <i *ngIf="displayInteractorOrganism"
    class="{{interactorOrganismIcon}}"
     title="{{interactorOrganism}}"></i>
  {{ interactorName }}
  <i *ngIf="displayInteractorType"
     class="{{interactorTypeIcon}}"
     title="{{interactorType}}"></i>
  <!-- Adding of the icons to access details of the interactor -->
  <a [routerLink]="['/complex/search']"
     [queryParams]="{query: interactorId, page: 1}"
     [fragment]="'view_complex_navigator'"
     target="_blank">
    <i class="icon icon-functional small" data-icon="1"
       title="Complexes containing this interactor"></i>
  </a>
  <cp-table-interactor-external-link
    [interactorId]="interactorId"
    [interactorType]="interactorType"
    [identifierLink]="identifierLink"
    [displayInteractorIdentifier]="displayInteractorIdentifier"
  ></cp-table-interactor-external-link>
</div>
```

And update the inputs of this component too
```
  @Input() identifierLink: string;
  @Input() displayInteractorOrganism: boolean;
  @Input() displayInteractorType: boolean;
  @Input() displayInteractorIdentifier: boolean;
```

In `table-interactor-column.component` I also added new inputs
```
  @Input() displayInteractorOrganism: boolean;
  @Input() displayInteractorType: boolean;
  @Input() displayInteractorIdentifier: boolean;
```
And in the HTML I commented out `cp-table-interactor-external-link` and update the 2 `cp-table-interactor-name` with the new inputs.
In `table-structure.component` I also added the new inputs
```
  @Input() displayInteractorOrganism: boolean;
  @Input() displayInteractorType: boolean;
  @Input() displayInteractorIdentifier: boolean;
```
And update the HTML to pass them to `cp-table-interactor-column`.

And finally, in `complex-navigator.component`, I updated `cp-table-structure` to pass the new inputs.
```
  <cp-table-structure [complexSearch]="complexSearch"
                      [interactors]="interactors"
                      [interactorsSorting]="interactorsSorting"
                      [displayInteractorOrganism]="displayInteractorOrganism"
                      [displayInteractorType]="displayInteractorType"
                      [displayInteractorIdentifier]="displayInteractorIdentifier">
  </cp-table-structure>
```

susiehgt committed:
Testing

susiehgt committed:
undoing of previous change
